File: Sources/YBarKit/Events/EventBus.swift
import Foundation

/// Named events with per-item subscription bitmasks (u64, sketchybar-compatible).
/// Built-ins occupy the low bits; custom events (`--add event`) append, optionally
/// bound to an NSDistributedNotificationCenter name whose userInfo is JSON-serialized
/// into $INFO — the cheap hook that makes Spotify-style integrations work.
@MainActor
public final class EventBus {
    public struct EventDefinition {
        public let name: String
        public let bit: UInt64
        public let notificationName: String?
    }

    public static let builtinEvents: [String] = [
        "front_app_switched", "space_change", "display_change",
        "system_woke", "system_will_sleep",
        "mouse.entered", "mouse.exited", "mouse.clicked", "mouse.scrolled",
        "volume_change", "power_source_change", "battery_change", "wifi_change",
        "system_stats", "mouse.exited.global",
    ]

    public private(set) var definitions: [EventDefinition] = []
    private var observers: [String: NSObjectProtocol] = [:]

    /// Wired by the daemon: run an item's script with the given environment.
    public var runItemScript: ((Item, [String: String]) -> Void)?
    /// Items provider (the store lives on BarManager).
    public var itemsProvider: (() -> [Item])?
    /// Lazy provider arming: called the first time any item subscribes to an event.
    public var onFirstSubscription: ((String) -> Void)?

    private var armedEvents: Set<String> = []

    public init() {
        for name in EventBus.builtinEvents {
            appendDefinition(name: name, notificationName: nil)
        }
    }

    public func reset() {
        for observer in observers.values {
            DistributedNotificationCenter.default().removeObserver(observer)
        }
        observers.removeAll()
        definitions.removeAll()
        armedEvents.removeAll()
        for name in EventBus.builtinEvents {
            appendDefinition(name: name, notificationName: nil)
        }
    }

    public func bit(for name: String) -> UInt64? {
        definitions.first { $0.name == name }?.bit
    }

    /// Register a custom event; no-op if it already exists. 64-event cap.
    @discardableResult
    public func addEvent(name: String, notificationName: String? = nil) -> String? {
        if bit(for: name) != nil { return nil }
        guard definitions.count < 64 else { return "[!] event limit reached (64)" }
        appendDefinition(name: name, notificationName: notificationName)
        if let notificationName {
            observeDistributedNotification(named: notificationName, eventName: name)
        }
        return nil
    }

    public func subscribe(item: Item, eventName: String) -> String? {
        guard let bit = bit(for: eventName) else {
            return "[!] unknown event: \(eventName)"
        }
        item.updateMask |= bit
        if !armedEvents.contains(eventName) {
            armedEvents.insert(eventName)
            onFirstSubscription?(eventName)
        }
        return nil
    }

    /// Fire an event to all subscribed items' scripts. The `updates` policy gates
    /// event-driven runs exactly like routine ones (sketchybar contract); mouse
    /// events and `--update` intentionally bypass it via their own paths.
    public func trigger(name: String, info: String = "", extraEnvironment: [String: String] = [:]) {
        guard let bit = bit(for: name), let items = itemsProvider?() else { return }
        for item in items where item.updateMask & bit != 0
            && (!item.script.isEmpty || item.hasLuaHandlers) {
            if item.updatePolicy == .off { continue }
            // when_shown gates on drawing (sketchybar semantics), NOT on content:
            // an empty item must still receive the update that populates it.
            if item.updatePolicy == .whenShown, !item.drawing { continue }
            // Contract variables are applied last: a `--trigger e NAME=x` payload
            // must never clobber the receiving item's identity.
            var environment = extraEnvironment
            environment["NAME"] = item.name
            environment["SENDER"] = name
            environment["INFO"] = info
            runItemScript?(item, environment)
        }
    }

    private func appendDefinition(name: String, notificationName: String?) {
        let bit: UInt64 = 1 << UInt64(definitions.count)
        definitions.append(EventDefinition(name: name, bit: bit, notificationName: notificationName))
    }

    private func observeDistributedNotification(named notificationName: String, eventName: String) {
        let observer = DistributedNotificationCenter.default().addObserver(
            forName: Notification.Name(notificationName),
            object: nil,
            queue: .main
        ) { notification in
            let info: String
            if let userInfo = notification.userInfo,
               let data = try? JSONSerialization.data(withJSONObject: userInfo, options: [.sortedKeys]),
               let json = String(data: data, encoding: .utf8) {
                info = json
            } else {
                info = ""
            }
            MainActor.assumeIsolated {
                // Re-resolve through the default center's main-queue delivery.
                NotificationBridge.shared.deliver(eventName: eventName, info: info)
            }
        }
        observers[notificationName] = observer
    }
}

/// Small indirection so distributed-notification closures don't retain the bus.
@MainActor
final class NotificationBridge {
    static let shared = NotificationBridge()
    weak var bus: EventBus?

    func deliver(eventName: String, info: String) {
        bus?.trigger(name: eventName, info: info)
    }
}

File: src/events/event_bus.cpp
#include "events/event_bus.h"

namespace ybar::events {

using ybar::model::Item;
using ybar::model::UpdatePolicy;

const std::vector<std::string>& EventBus::builtinEvents() {
    // Exact declaration order fixes bit values — contract (spec 3.4).
    static const std::vector<std::string> events = {
        "front_app_switched", "space_change", "display_change", "system_woke",
        "system_will_sleep", "mouse.entered", "mouse.exited", "mouse.clicked",
        "mouse.scrolled", "volume_change", "power_source_change", "battery_change",
        "wifi_change", "system_stats", "mouse.exited.global", "mouse.entered.global",
        "modifier_change", "app_launched", "app_terminated", "media_change",
    };
    return events;
}

EventBus::EventBus() { reset(); }

void EventBus::reset() {
    definitions_.clear();
    armed_.clear();
    for (const auto& name : builtinEvents()) {
        definitions_.push_back({name, 1ull << definitions_.size(), {}});
        armed_.push_back(false);
    }
}

std::optional<std::string> EventBus::addEvent(const std::string& name,
                                              const std::string& notification) {
    if (knownEvent(name)) return std::nullopt; // duplicate add is a silent no-op
    if (definitions_.size() >= kMaxEvents) return "[!] event limit reached (64)";
    definitions_.push_back({name, 1ull << definitions_.size(), notification});
    armed_.push_back(false);
    return std::nullopt;
}

bool EventBus::knownEvent(std::string_view name) const {
    return eventBit(name).has_value();
}

std::optional<std::uint64_t> EventBus::eventBit(std::string_view name) const {
    for (const auto& def : definitions_)
        if (def.name == name) return def.bit;
    return std::nullopt;
}

bool EventBus::subscribe(Item& item, std::string_view eventName) {
    for (std::size_t i = 0; i < definitions_.size(); ++i) {
        if (definitions_[i].name != eventName) continue;
        item.updateMask |= definitions_[i].bit;
        if (!armed_[i] && onFirstSubscription) {
            armed_[i] = true;
            onFirstSubscription(definitions_[i].name);
        }
        return true;
    }
    return false;
}

namespace {

Environment merged(const Environment& extra, const Item& item, std::string_view sender,
                   const std::string& info) {
    Environment env = extra;
    // Applied LAST — a --trigger payload can never clobber these.
    env["NAME"] = item.name;
    env["SENDER"] = std::string(sender);
    env["INFO"] = info;
    return env;
}

} // namespace

void EventBus::trigger(std::string_view eventName, const std::string& info,
                       const Environment& extraEnvironment) {
    const auto bit = eventBit(eventName);
    if (!bit || !itemsProvider || !runItemScript) return;
    for (auto* item : itemsProvider()) {
        if (!(item->updateMask & *bit)) continue;
        if (item->script.empty() && !item->hasLuaHandlers) continue;
        if (item->updatePolicy == UpdatePolicy::Off) continue;
        if (item->updatePolicy == UpdatePolicy::WhenShown && !item->drawing) continue;
        runItemScript(*item, merged(extraEnvironment, *item, eventName, info));
    }
}

void EventBus::triggerTargeted(Item& item, std::string_view eventName, const std::string& info,
                               const Environment& extraEnvironment) {
    // Mouse events bypass the updates-POLICY gate, but the subscription mask
    // still governs delivery (spec 3.4) — otherwise every scripted item fires
    // on any hover/click.
    if (!runItemScript) return;
    const auto bit = eventBit(eventName);
    if (!bit || !(item.updateMask & *bit)) return;
    if (item.script.empty() && !item.hasLuaHandlers) return;
    runItemScript(item, merged(extraEnvironment, item, eventName, info));
}

} // namespace ybar::events
